libpod/image/utils.go

```diff
@@ -20,7 +20,32 @@ import (
 // a match on name:tag
 func findImageInRepotags(search imageParts, images []*Image) (*storage.Image, error) {
 	_, searchName, searchSuspiciousTagValueForSearch := search.suspiciousRefNameTagValuesForSearch()
-	var results []*storage.Image
+	var (
+		results []*storage.Image
+		foundRW bool
+	)
+
+	// This function prefers read/write images over read/only images, if they match
+	appendImage := func(results []*storage.Image, image *Image) []*storage.Image {
+		if foundRW {
+			// If readwrite image exists ignore readonly images
+			if image.IsReadOnly() {
```

Comment:
So if, searching for `alpine`, there is a read-only `example.com/alpine`, and a read-write `example.net/alpine`, this will silently ignore the ambiguity and only return the read-write one. That doesn’t look right to me.

BTW `results[]` can only legitimately contain 0 or 1 element, so eliminating the array and checking for equality (or rather, image ID identity? I’m not at all sure) should be fairly easy.

Reply:
I believe results could container more then 1, if you had X additional stores, each one with a readonly  alpine, then we would have X results.

Reply:
On your first comment, I would say that I believe the goal of the search is to find the image that the application should use.  In the case of multiple images with only one Read/Write, then the Read/Write image should be used.  If there are multiple read/only or multiple read/writes, then we get the ambiguity.

Reply:
I meant that the _return value_ of the function is at most one value, or the function returns an error. So collecting more than one already guarantees that there will be an error.

WRT preferring read-write:
- AFAICT after a fair bit of looking, the purpose of this heuristic search is completely undocumented. It’s a bit tempting to guess that it only exists as a workaround for some inaccurate/incorrect search that used to exist elsewhere (in /usr/bin/atomic quite likely). So I’m very reluctant to add any more rules to the heuristic (although given I have no idea why this exists, I should have very little say).
- The original report already starts with `podman image exists`, which really does not care about read-write/read-only one way or the other. `podman run` can, AFAIK, also work just fine with read-only images (wasn’t that the whole point of adding additional read-only stores?). I’d even argue that _most_ users don’t care one bit whether the store is in a read-write store or not.

Reply:
I'm not entirely following this conversation, so maybe my question is obvious/inapplicable: what is the intended purpose of `podman exists`? From my perspective, `podman exists foo` is a tool that tells me whether or not `podman run foo` will work. (Or `podman inspect foo`, or so on). Does this approach preserve that behavior?

Reply:
To the extent both use `getLocalImage`, which I think is the case, both will be affected equally by whatever change is made.

Reply:
Ok I simplified the function to return a read/write image, if it finds one, otherwise it will return the read/only image, if it exists.  Will return error if no match.

Ignores the case where there are multiple R/O images with the same name.

Reply:
It also ignores the case where there are multiple ambiguous R/O images with different names, and multiple ambiguous R/W images with different names, as well as the earlier-discussed R/O vs. R/W name ambiguity, and makes `ErrMultipleImages` impossible to reach without fixing up consumers that expect it.

The simplification of the check is nice, though.

---

I was suggesting
```go
var res *storage.Image
…
if image /* matches */ {
    if res != nil && { // is ID the right thing to check across stores???
        return define.ErrMultipleImages
    }
    if res == nil || res.IsReadOnly { // image may be RO or RW
           res = image
    }
````
+ more or less the current after-the-loop code checking `res`.

---

+ the “podman image exists” check for `ErrMultipleImages` needs either fixing or removing (I can’t immediately remember which one), either way.

Reply:
(Updated the suggestion with an ID condition I lost when copy&pasting.)